### DISCUSSION TITLE
Difficulty synthesizing a function that checks if a list is nil

### DISCUSSION THREAD

**Discussion by langston-barrett** [cvc5/cvc5#10160] [Q&A] [answered]
I tried to use CVC5 to solve the following SyGuS query:
```
(set-logic DTLIA)
(declare-datatype List
  ((nil)
   (cons (head Int) (tail List))))
(synth-fun f ((x List)) Bool
  ((b Bool)
   (c Int)
   (l List))
  ((b Bool (true false ((_ is nil) l) ((_ is cons) l)))
   (c Int (0 1))
   (l List (nil (cons c l)))))
(constraint (= (f nil) true))
(constraint (forall ((i Int) (l List)) (= (f (cons i l)) false)))
(check-synth)
```
I expected output like this:
```
(
(define-fun f ((x List)) Bool ((_ is nil) x))
)
```
but instead, the solver does not produce any response.

I tried the command-line options mentioned [here](https://github.com/cvc5/cvc5/discussions/9201), as well as replacing the `forall` with `declare-var`, but neither seemed to make a difference. Am I phrasing this problem incorrectly?

Many thanks for your time!

**Comment by ajreynol** [comment, marked as answer]
Hello,
The main issue is that your grammar does not allow `x` as an argument to `(_ is nil)`.  You should update the grammar to include it.
You should also use `declare-var` as opposed to `forall`.

cvc5 solves this problem quickly with the intended solution:
```
(set-logic DTLIA)
(declare-datatype List
  ((nil)
   (cons (head Int) (tail List))))
(synth-fun f ((x List)) Bool
  ((b Bool)
   (c Int)
   (l List))
  ((b Bool (true false ((_ is nil) l) ((_ is cons) l)))
   (c Int (0 1))
   (l List (x nil (cons c l)))))
(declare-var i Int)
(declare-var l List)
(constraint (= (f nil) true))
(constraint (= (f (cons i l)) false))
(check-synth)
```

Ideally cvc5 should report `infeasible` for your original problem, although unfortunately we don't use any sophisticated techniques to determine that no solution exists.  This is something we are planning to improve our support for in the future.

Hope this helps,
Andy
